### PR TITLE
Liquidity protection store upgrade phase 1

### PR DESCRIPTION
**Pull request by barakman**
1. Implement the `SystemStore` contract with functions for seeding the system balances and the minted token amounts.
2. Implement a script for migrating the system balances from the `Store` contract and the minted token amounts from the `Settings` contract into the `SystemStore` contract.
3. Implement in the main contract a function for withdrawing all tokens from the `Store` contract into a `TokenHolder` contract held in the main contract.
4. Add tests for the `SystemStore` contract.
5. Fix the deployment script according to all contract changes.

### CODE REVIEW

File: solidity/contracts/liquidity-protection/LiquidityProtection.sol [hunk 1]
```diff
@@ -74,6 +74,8 @@ contract LiquidityProtection is ILiquidityProtection, TokenHandler, Utils, Owned
     ILiquidityProtectionSettings public immutable override settings;
     ILiquidityProtectionStore public immutable override store;
     ILiquidityProtectionStats public immutable override stats;
+    ILiquidityProtectionSystemStore public immutable override systemStore;
+    ITokenHolder public immutable override tokenHolder;
```
Comment (yudilevi) [inline solidity/contracts/liquidity-protection/LiquidityProtection.sol:78]: `tokenHolder` -> `wallet`

File: solidity/contracts/liquidity-protection/LiquidityProtection.sol [hunk 2]
```diff
@@ -205,6 +196,42 @@ contract LiquidityProtection is ILiquidityProtection, TokenHandler, Utils, Owned
         store.acceptOwnership();
     }
 
+    /**
+     * @dev transfers the ownership of the token holder
+     * can only be called by the contract owner
+     *
+     * @param _newOwner    the new owner of the token holder
+     */
+    function transferTokenHolderOwnership(address _newOwner) external ownerOnly {
```
Comment (yudilevi) [inline solidity/contracts/liquidity-protection/LiquidityProtection.sol:205]: `transferWallet....` etc. everywhere

File: solidity/contracts/liquidity-protection/LiquidityProtection.sol [hunk 3]
```diff
@@ -205,6 +196,42 @@ contract LiquidityProtection is ILiquidityProtection, TokenHandler, Utils, Owned
...
+    /**
+     * @dev accepts the ownership of the token holder
+     * can only be called by the contract owner
+     */
+    function acceptTokenHolderOwnership() external ownerOnly {
+        tokenHolder.acceptOwnership();
+    }
+
+    /**
+     * @dev transfer all equity from the store to the token holder
```
Comment (yudilevi) [inline solidity/contracts/liquidity-protection/LiquidityProtection.sol:218]: migrates all funds from the store to the wallet

File: solidity/contracts/liquidity-protection/LiquidityProtection.sol [hunk 4]
```diff
@@ -205,6 +196,42 @@ contract LiquidityProtection is ILiquidityProtection, TokenHandler, Utils, Owned
...
+     */
+    function withdrawTokens() external {
```
Comment (yudilevi) [inline solidity/contracts/liquidity-protection/LiquidityProtection.sol:220]: migrateFunds

File: solidity/contracts/liquidity-protection/LiquidityProtectionStats.sol [hunk 5]
```diff
@@ -55,7 +55,7 @@ contract LiquidityProtectionStats is ILiquidityProtectionStats, AccessControl {
 
     /**
      * @dev increases the total amounts
-     * can only be executed only by an owner
+     * can be executed only by an owner
```
Comment (yudilevi) [inline solidity/contracts/liquidity-protection/LiquidityProtectionStats.sol:58]: previous version is cleaner

File: solidity/contracts/liquidity-protection/LiquidityProtectionSystemStore.sol [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+import "../utility/Utils.sol";
+
+/**
+ * @dev This contract aggregates the system balances of the liquidity protection mechanism.
+ */
+contract LiquidityProtectionSystemStore is ILiquidityProtectionSystemStore, AccessControl, Utils {
+    using SafeMath for uint256;
+
+    bytes32 public constant ROLE_SUPERVISOR = keccak256("ROLE_SUPERVISOR");
+    bytes32 public constant ROLE_SEEDER = keccak256("ROLE_SEEDER");
```
Comment (yudilevi) [inline solidity/contracts/liquidity-protection/LiquidityProtectionSystemStore.sol:18]: we don't really need a seeder role here, we can do the same thing we do for the wallet, just a function that migrates everything in one shot.
one function for the limits, another for the minted amounts.

**Review by yudilevi — CHANGES_REQUESTED** [review]

File: solidity/contracts/liquidity-protection/LiquidityProtection.sol (hunk 4 above)
Comment (barakman) [reply solidity/contracts/liquidity-protection/LiquidityProtection.sol]: Fixed.

File: solidity/contracts/liquidity-protection/LiquidityProtectionStats.sol (hunk 5 above)
Comment (barakman) [reply solidity/contracts/liquidity-protection/LiquidityProtectionStats.sol:58]: Previous version is wrong (an extra 'only').

File: solidity/contracts/liquidity-protection/LiquidityProtectionSystemStore.sol (hunk 6 above)
Comment (barakman) [reply solidity/contracts/liquidity-protection/LiquidityProtectionSystemStore.sol]: Fixed.

File: solidity/contracts/liquidity-protection/LiquidityProtection.sol (hunk 3 above)
Comment (barakman) [reply solidity/contracts/liquidity-protection/LiquidityProtection.sol]: Fixed.

File: solidity/contracts/liquidity-protection/LiquidityProtection.sol (hunk 2 above)
Comment (barakman) [reply solidity/contracts/liquidity-protection/LiquidityProtection.sol]: Fixed.

File: solidity/contracts/liquidity-protection/LiquidityProtection.sol (hunk 1 above)
Comment (barakman) [reply solidity/contracts/liquidity-protection/LiquidityProtection.sol]: Fixed.

**Review by yudilevi — CHANGES_REQUESTED** [review]
I believe the test deployment script wasn't updated

**Comment by barakman** [comment]
> I believe the test deployment script wasn't updated

What makes you think that?

**Review by yudilevi — APPROVED** [review]

**Pull request merged**